sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeExpressions.scala

```diff
@@ -589,22 +589,43 @@ case class DateFormatClass(left: Expression, right: Expression, timeZoneId: Opti
 
   override def inputTypes: Seq[AbstractDataType] = Seq(TimestampType, StringType)
 
-  override def withTimeZone(timeZoneId: String): TimeZoneAwareExpression =
+  var formatter: Option[TimestampFormatter] = None
```

Comment:
can we do it like this?
```

  lazy val formatter = if (right.foldable) {
    val fmt = right.eval().toString
    val formatter = TimestampFormatter(fmt, zoneId, Locale.US)
    (fmt: Any) => formatter
  } else {
    (fmt: Any) => TimestampFormatter(fmt.toString, zoneId, Locale.US)
  }

  override protected def nullSafeEval(timestamp: Any, format: Any): Any = {
    UTF8String.fromString(formatter(format).format(timestamp.asInstanceOf[Long]))
  }
```

Reply:
In the case of `doGenCode`, what should I pass to `addReferenceObj()`? And call this lambda `formatter` per each value? Could you explain, please, why your code is better.

Reply:
Just to avoid `var` in the earlier commit. But, in the latest commit. `var`'s gone away, so the current one looks ok to me.